Userland/Libraries/LibCore/MemoryStream.cpp

```diff
@@ -125,4 +126,120 @@ size_t FixedMemoryStream::remaining() const
     return m_bytes.size() - m_offset;
 }
 
+ErrorOr<Bytes> AllocatingMemoryStream::read(Bytes bytes)
+{
+    size_t read_bytes = 0;
+
+    while (read_bytes < bytes.size()) {
+        VERIFY(m_write_offset >= m_read_offset);
+
+        auto range = TRY(next_read_range());
+        if (range.size() == 0)
+            break;
+
+        auto copied_bytes = range.copy_trimmed_to(bytes.slice(read_bytes));
+
+        read_bytes += copied_bytes;
+        m_read_offset += copied_bytes;
+    }
+
+    cleanup_unused_chunks();
+
+    return bytes.trim(read_bytes);
+}
+
+ErrorOr<size_t> AllocatingMemoryStream::write(ReadonlyBytes bytes)
+{
+    size_t written_bytes = 0;
+
+    while (written_bytes < bytes.size()) {
+        VERIFY(m_write_offset >= m_read_offset);
+
+        auto range = TRY(next_write_range());
+
+        auto copied_bytes = bytes.slice(written_bytes).copy_trimmed_to(range);
+
+        written_bytes += copied_bytes;
+        m_write_offset += copied_bytes;
+    }
+
+    return written_bytes;
+}
+
+ErrorOr<void> AllocatingMemoryStream::discard(size_t count)
+{
+    VERIFY(m_write_offset >= m_read_offset);
+
+    if (count > size())
+        return Error::from_string_literal("Number of discarded bytes is higher than the number of allocated bytes");
+
+    m_read_offset += count;
+
+    cleanup_unused_chunks();
+
+    return {};
+}
+
+bool AllocatingMemoryStream::is_eof() const
+{
+    return size() == 0;
+}
+
+size_t AllocatingMemoryStream::size() const
+{
+    return m_write_offset - m_read_offset;
+}
+
+bool AllocatingMemoryStream::is_open() const
+{
+    return true;
+}
+
+void AllocatingMemoryStream::close()
+{
+}
+
+ErrorOr<ReadonlyBytes> AllocatingMemoryStream::next_read_range()
+{
+    VERIFY(m_write_offset >= m_read_offset);
+
+    size_t const chunk_index = m_read_offset / chunk_size;
+    size_t const chunk_offset = m_read_offset % chunk_size;
+    size_t const read_size = min(chunk_size - m_read_offset % chunk_size, m_write_offset - m_read_offset);
+
+    if (read_size == 0)
+        return ReadonlyBytes { static_cast<u8*>(nullptr), 0 };
+
+    VERIFY(m_chunks.size() > chunk_index);
+
+    return ReadonlyBytes { m_chunks[chunk_index].data() + chunk_offset, read_size };
+}
+
+ErrorOr<Bytes> AllocatingMemoryStream::next_write_range()
+{
+    VERIFY(m_write_offset >= m_read_offset);
+
+    size_t const chunk_index = m_write_offset / chunk_size;
+    size_t const chunk_offset = m_write_offset % chunk_size;
+    size_t const write_size = chunk_size - m_write_offset % chunk_size;
+
+    if (chunk_index >= m_chunks.size())
+        m_chunks.append(TRY(ByteBuffer::create_uninitialized(chunk_size)));
+
+    return Bytes { m_chunks[chunk_index].data() + chunk_offset, write_size };
+}
+
+void AllocatingMemoryStream::cleanup_unused_chunks()
+{
+    while (m_read_offset >= chunk_size) {
+        VERIFY(m_write_offset >= m_read_offset);
+
+        auto buffer = m_chunks.take_first();
```

Comment:
This is pretty expensive since we have to move every chunk for each `take_first` could we compute how many we move first then move all of them at once?

Reply:
In theory we could, but I don't think that we have a `take_multiple` in `Vector` as of now, and adding one is probably outside of the scope of this PR (and not particularly my area of C++-expertise). I'd be fine with adding it as a `FIXME` for the future though.

However, we should be working with non-inline storage here, right? So moving around those ByteBuffers wouldn't be _that_ expensive.

Reply:
What I could instead do right now is to make those `ByteBuffer<0>`, which shrinks the data area from 32 bytes down to a pointer and a capacity (`m_size` and `m_inline` are common between inline and outline buffer mode).